README.md
# `prometheus_aggregator`

A Ruby client for https://github.com/peterbourgon/prometheus-aggregator.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'prometheus_aggregator'
```

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
## 1.0.0 - 2018-11-07
- Support gauges

## 0.2.0 - 2018-11-07
- Add Faraday middleware

## 0.1.1 - 2018-11-01
- Make queue capacity configurable

## 0.1.0 - 2018-11-01
- Initial release, including thread/fork-safe client and Rack middleware


test/client_test.rb
# frozen_string_literal: true

require "excon"
require "test_helper"
require "prometheus_aggregator/client"

class ClientTest < Minitest::Test
  def setup
    AggregatorServer.start
  end

  def teardown
    AggregatorServer.stop
  end

  def test_counters_work_correctly
    client = AggregatorServer.client
    client.counter(name: "test_counter", value: 1, help: "Help text")
    client.counter(name: "test_counter", value: 1, help: "Help text")
    sleep 0.1

    scrape_result = AggregatorServer.scrape_metrics
    assert_includes scrape_result, "# HELP test_counter Help text\n"
    assert_includes scrape_result, "# TYPE test_counter counter\n"
    assert_includes scrape_result, "test_counter{} 2.0"

    client.stop
  end

  def test_gauges_work_correctly
    client = AggregatorServer.client
    client.gauge(name: "test_gauge", value: 1, help: "Help text")
    client.gauge(name: "test_gauge", value: 2, help: "Help text")
    sleep 0.1

    scrape_result = AggregatorServer.scrape_metrics
    assert_includes scrape_result, "# HELP test_gauge Help text\n"
    assert_includes scrape_result, "# TYPE test_gauge gauge\n"
    assert_includes scrape_result, "test_gauge{} 2.0"

    client.stop
  end

  def test_histograms_work_correctly
    client = AggregatorServer.client
    client.histogram(name: "test_histogram", value: 0.9, help: "Help text")
    sleep 0.1

    scrape_result = AggregatorServer.scrape_metrics
    assert_includes scrape_result, "# HELP test_histogram Help text\n"
    assert_includes scrape_result, "# TYPE test_histogram histogram\n"
    assert_includes scrape_result, "test_histogram_bucket{le=\"0.01\"} 0\n"
    assert_includes scrape_result, "test_histogram_bucket{le=\"1\"} 1\n"

    client.stop
  end

  def test_handles_disconnections
    client = AggregatorServer.client

    client.counter(name: "test_counter_1", value: 1, help: "Help text")

    AggregatorServer.stop

    # Shouldn't blow up
    client.counter(name: "test_counter_1", value: 1, help: "Help text")

    client.stop
  end

  def test_handles_reconnections
    client = AggregatorServer.client

    client.counter(name: "test_counter_1", value: 1, help: "Help text")

    AggregatorServer.stop
    AggregatorServer.start

    # Connection retries happen every 0.1 seconds
    sleep 0.2
    client.counter(name: "test_counter_2", value: 1, help: "Help text")
    sleep 0.1

    assert_includes AggregatorServer.scrape_metrics, "test_counter_2{} 1.0"

    client.stop
  end

  def test_records_are_buffered
    AggregatorServer.stop

    client = AggregatorServer.client

    5.times do
      client.counter(name: "test_counter", value: 1, help: "Help text")
    end

    AggregatorServer.start

    # Automatic reconnect attempts happen every 0.1 seconds
    sleep 0.2

    assert_includes AggregatorServer.scrape_metrics, "test_counter{} 5.0"

    client.stop
  end

  def test_that_stale_events_are_dropped
    AggregatorServer.stop

    client = AggregatorServer.client(staleness_threshold: 1.0)

    client.counter(name: "test_counter", value: 1, help: "Help text")
    client.counter(name: "test_counter", value: 1, help: "Help text")

    sleep 1.0

    client.counter(name: "test_counter", value: 1, help: "Help text")
    client.counter(name: "test_counter", value: 1, help: "Help text")

    AggregatorServer.start

    sleep 0.2

    assert_includes AggregatorServer.scrape_metrics, "test_counter{} 2.0"

    client.stop
  end
end


test/test_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "prometheus_aggregator"

require "minitest/autorun"
require "excon"

PrometheusAggregator.logger = Logger.new(File.open("/dev/null", "w"))

module AggregatorServer
  LISTEN_PORT = 18_191
  SCRAPE_PORT = 18_192

  def self.scrape_metrics
    Excon.get("http://127.0.0.1:#{SCRAPE_PORT}/metrics").body
  end

  def self.client(opts = {})
    opts = { connection_retry_interval: 0.1 }.merge(opts)
    PrometheusAggregator::Client.new("127.0.0.1", LISTEN_PORT, opts)
  end

  def self.start
    download_binary unless File.exist?(binary_path)

    listen_addr = "tcp://127.0.0.1:#{LISTEN_PORT}"
    scrape_url = "tcp://127.0.0.1:#{SCRAPE_PORT}/metrics"
    args = ["-socket", listen_addr, "-prometheus", scrape_url]
    @pid = IO.popen(([binary_path] + args).join(" ")).pid

    wait_for_server
  end

  def self.stop(signal: :SIGTERM)
    start_time = Time.now

    begin
      Process.kill(signal, @pid)

      loop do
        sleep 0.1
        Process.getpgid(@pid)
        break if Time.now > start_time + 2.0
      end
    rescue Errno::ESRCH
      # Process is dead
      return
    end

    stop(:SIGKILL)
  end

  def self.wait_for_server
    retries = 0
    begin
      socket = TCPSocket.new("127.0.0.1", LISTEN_PORT)
      socket.close
    rescue Errno::ECONNREFUSED => err
      raise err if retries >= 20

      sleep 0.1
      retries += 1
      retry
    end
  end

  def self.binary_path
    File.join(binary_dir, "prometheus-aggregator")
  end

  def self.binary_dir
    File.join(File.expand_path("..", __dir__), "tmp")
  end

  def self.download_binary
    puts " => Downloading prometheus-aggregator binary"
    repo = "peterbourgon/prometheus-aggregator"
    releases_url = "https://api.github.com/repos/#{repo}/releases/latest"
    response = JSON.parse(Excon.get(releases_url).body)
    asset = response["assets"].find { |a| a["name"].include?(platform) }

    download_url = Excon.get(asset["browser_download_url"]).headers["Location"]
    response = Excon.get(download_url, omit_default_port: true)
    FileUtils.mkdir_p(binary_dir)
    File.open(binary_path, "wb", 0o755) { |f| f.write(response.body) }
  end

  def self.platform
    case RbConfig::CONFIG["arch"]
    when /linux/ then "linux"
    when /darwin/ then "darwin"
    else raise "Invalid platform #{RbConfig::CONFIG['arch']}"
    end
  end
end


test/exporter_test.rb
# frozen_string_literal: true

require "test_helper"
require "prometheus_aggregator/client"

class ClientTest < Minitest::Test
  def test_capacity_is_enforced
    exporter = PrometheusAggregator::Exporter.new("localhost", 1234)
    105.times { exporter.enqueue(:foo) }

    capacity = PrometheusAggregator::Exporter::QUEUE_CAPACITY
    assert_equal capacity, exporter.backlog
  end
end


lib/prometheus_aggregator/exporter.rb
# frozen_string_literal: true

require "openssl"
require "json"
require "socket"
require "net_tcp_client"
require "prometheus_aggregator"

module PrometheusAggregator
  class Exporter
    CONNECTION_RETRY_INTERVAL = 1.0
    QUEUE_CAPACITY = 100
    LOOP_INTERVAL = 0.01
    STALENESS_THRESHOLD = 5.0

    def initialize(host, port, opts = {})
      @host = host
      @port = port
      @ssl_params = opts[:ssl_params]
      @connection_retry_interval =
        opts[:connection_retry_interval] || CONNECTION_RETRY_INTERVAL
      @queue_capacity = opts[:queue_capacity] || QUEUE_CAPACITY
      @staleness_threshold = opts[:staleness_threshold] || STALENESS_THRESHOLD
      @registered = {}

      @stop = false
      @pid = nil
      @mutex = Mutex.new
      @queue = []
      @socket = nil
      @thread = nil
    end

    def enqueue(record)
      @mutex.synchronize do
        if Process.pid != @pid
          PrometheusAggregator.logger.info("Pid mismatch, spawning new thread")
          @thread&.kill
          @thread = nil
        end

        if @thread.nil?
          @thread = Thread.new { write_loop }
          @pid = Process.pid
        end

        @queue << [Time.now, record]
        @queue.shift while @queue.length > QUEUE_CAPACITY
      end
    end

    def backlog
      @queue.length
    end

    def stop
      @stop = true
    end

    private

    def write_loop
      loop do
        break if @stop

        connect unless connection_ok?
        unless connection_ok?
          PrometheusAggregator.logger.warn(
            "Not connected to prometheus agggregator (#{@host}:#{@port})"
          )

          sleep(@connection_retry_interval)
          next
        end

        event = @mutex.synchronize do
          @queue.shift while !@queue.empty? && stale?(@queue.first)
          @queue.shift
        end

        if event.nil?
          sleep(LOOP_INTERVAL)
          next
        end

        record = event[1]
        register(record)
        emit_value(record)
      end
    end

    def stale?(record)
      record[0] < Time.now - @staleness_threshold
    end

    def connection_ok?
      !@socket.nil? && @socket.alive?
    end

    def connect
      @registered = {}

      @socket&.close

      @socket = Net::TCPClient.new(
        server: "#{@host}:#{@port}",
        ssl: @ssl_params,
        connect_timeout: 3.0,
        write_timeout: 3.0,
        read_timeout: 3.0,
        connect_retry_count: 0
      )
    rescue Net::TCPClient::ConnectionFailure => err
      PrometheusAggregator.logger.debug(err)
      @socket&.close
      @socket = nil
    end

    def register(record)
      declaration = record.reject { |k| k == :value }
      json_declaration = JSON.fast_generate(declaration)
      return if @registered[json_declaration]

      send_line(json_declaration)
      @registered[json_declaration] = true
    end

    def emit_value(record)
      send_line(JSON.fast_generate(record.slice(:name, :value, :labels)))
    end

    def send_line(line)
      @socket.write(line + "\n")
    rescue Net::TCPClient::ConnectionFailure => err
      PrometheusAggregator.logger.debug(err)
      @socket&.close
      @socket = nil
    end
  end
end


lib/prometheus_aggregator/rack_middleware.rb
# frozen_string_literal: true

require "prometheus_aggregator"

module PrometheusAggregator
  class RackMiddleware
    def initialize(app, options = {})
      @app = app
      @client = options[:client]
      raise ArgumentError, ":client option is required" unless @client
    end

    def call(env)
      start_time = Time.now
      response = @app.call(env)
      duration = Time.now - start_time

      begin
        @client.counter(
          name: "http_server_requests_total",
          help: "The total number of HTTP requests handled by the Rack app",
          value: 1,
          labels: labels(env).merge(code: response.first.to_s)
        )

        @client.histogram(
          name: "http_server_request_duration_seconds",
          help: "The HTTP response duration of the Rack application",
          value: duration,
          labels: labels(env)
        )
      rescue => err # rubocop:disable Style/RescueStandardError
        # Let's be ultra defensive. Metrics should never break the app.
        PrometheusAggregator.logger.error("RackMiddleware: #{err}")
      end

      response
    end

    def labels(env)
      {
        method: env["REQUEST_METHOD"].downcase,
        path: clean_path(env["PATH_INFO"])
      }
    end

    def clean_path(path)
      path.gsub(%r{/\d+/}, "/:id/").gsub(%r{/\d+$}, "/:id")
    end
  end
end


lib/prometheus_aggregator/faraday_middleware.rb
# frozen_string_literal: true

require "faraday"
require "prometheus_aggregator"

module PrometheusAggregator
  class FaradayMiddleware < Faraday::Middleware
    def initialize(app, options = {})
      super(app)
      @client = options[:client]
      raise ArgumentError, ":client option is required" unless @client
    end

    def call(request_env)
      start_time = Time.now
      @app.call(request_env).on_complete do |response_env|
        duration = Time.now - start_time
        status_code = response_env[:status].to_s

        begin
          @client.counter(
            name: "http_client_requests_total",
            help: "The total number of HTTP requests sent by the client",
            value: 1,
            labels: labels(response_env).merge(code: status_code)
          )

          @client.histogram(
            name: "http_client_request_duration_seconds",
            help: "The HTTP response duration",
            value: duration,
            labels: labels(response_env)
          )
        rescue => err # rubocop:disable Style/RescueStandardError
          # Let's be ultra defensive. Metrics should never break the app.
          PrometheusAggregator.logger.error("FaradayMiddleware: #{err}")
        end
      end
    end

    private

    def labels(response_env)
      {
        method: response_env[:method].to_s,
        host: response_env[:url].host
      }
    end
  end
end


lib/prometheus_aggregator/client.rb
# frozen_string_literal: true

require "prometheus_aggregator/exporter"

module PrometheusAggregator
  class Client
    DEFAULT_BUCKETS = [0.01, 0.05, 0.1, 0.2, 0.5, 1, 2, 5, 10].freeze

    def initialize(host, port, opts = {})
      @default_labels = opts.delete(:default_labels) || {}
      @exporter = Exporter.new(host, port, opts)
    end

    def counter(name:, value:, help:, labels: {})
      @exporter.enqueue(
        type: "counter",
        name: name,
        value: value,
        help: help,
        labels: @default_labels.dup.merge(labels)
      )
    end

    def gauge(name:, value:, help:, labels: {})
      @exporter.enqueue(
        type: "gauge",
        name: name,
        value: value,
        help: help,
        labels: @default_labels.dup.merge(labels)
      )
    end

    def histogram(name:, value:, help:, buckets: DEFAULT_BUCKETS, labels: {})
      @exporter.enqueue(
        type: "histogram",
        name: name,
        value: value,
        help: help,
        buckets: buckets,
        labels: @default_labels.dup.merge(labels)
      )
    end

    def stop
      @exporter.stop
    end
  end
end
